### PR TITLE
SqlClient minor memory improvements

### PR DESCRIPTION
**Pull request by Wraith2**
Some work was done in https://github.com/dotnet/corefx/pull/34047 to limit allocation of closures and move to explicit state passing for continuations. This PR continues/finishes that work and cleans up some of the async code to be more readable.

AsyncHelper had a few methods which took unused parameters, either internal or external connections, with names that indicated those connections were going to be closed. The parameters were never used. I judged these to be dead code and eliminated the parameters and all call sites using them.

the function ContinueTaskPend was removed as dead code with no callers.

I recommend looking at the changes side by side, the inline diff makes them look confusing

/cc @saurabh500 @AfsanehR @keeratsingh

### CODE REVIEW

**Comment by krwq** [comment]
test Windows x86 Release Build

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs [hunk 1]
```diff
@@ -27,9 +27,22 @@ internal static Task CreateContinuationTask(Task task, Action onSuccess, SqlInte
             else
             {
                 TaskCompletionSource<object> completion = new TaskCompletionSource<object>();
-                ContinueTask(task, completion,
-                    () => { onSuccess(); completion.SetResult(null); },
-                    connectionToDoom, onFailure);
+                ContinueTaskWithState(task, completion,
+                    state: Tuple.Create(onSuccess, onFailure,completion),
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs]: Nit : Space after `onFailure, `

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs [hunk 2]
```diff
@@ -7522,17 +7522,18 @@ internal Task TdsExecuteSQLBatch(string text, int timeout, SqlNotificationReques
...
 
                                     // Take care of releasing the locks
                                     if (releaseConnectionLock)
                                     {
-                                        task.ContinueWith(_ =>
-                                        {
-                                            _connHandler._parserLock.Release();
-                                        }, TaskScheduler.Default);
+                                        task.ContinueWith(
+                                            (_,state) => ((SqlInternalConnectionTds)state)._parserLock.Release(),
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]:
```suggestion
                                            (_, state) => ((SqlInternalConnectionTds)state)._parserLock.Release(),
```

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj [hunk 3]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: Remove this file from the change.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs [hunk 4]
```diff
@@ -3118,8 +3118,7 @@ private Task WriteBytes(ReadOnlySpan<byte> b, int len, int offsetBuffer, bool ca
         private void WriteBytesSetupContinuation(byte[] array, int len, TaskCompletionSource<object> completion, int offset, Task packetTask)
         {
             AsyncHelper.ContinueTask(packetTask, completion,
-                () => WriteBytes(ReadOnlySpan<byte>.Empty, len: len, offsetBuffer: offset, canAccumulate: false, completion: completion, array),
-                connectionToDoom: _parser.Connection
+                () => WriteBytes(ReadOnlySpan<byte>.Empty, len: len, offsetBuffer: offset, canAccumulate: false, completion: completion, array)
```
Comment (saurabh500) [inline src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs]: Nit: Consistency on the usage of the function? 
```suggestion
                onSuccess: () => WriteBytes(ReadOnlySpan<byte>.Empty, len: len, offsetBuffer: offset, canAccumulate: false, completion: completion, array)
```

File: src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs (hunk 1 above)
Comment (saurabh500) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/SqlUtil.cs]:
```suggestion
                    state: Tuple.Create(onSuccess, onFailure, completion),
```

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs (hunk 4 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserStateObject.cs]: Done.

File: src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs (hunk 2 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParser.cs]: Done.

File: src/System.Data.SqlClient/src/System.Data.SqlClient.csproj (hunk 3 above)
Comment (Wraith2) [reply src/System.Data.SqlClient/src/System.Data.SqlClient.csproj]: I've reverted the file to a version before the encoding got changed, hopefully that removes it from the changeset.

**Comment by Wraith2** [comment]
@dotnet-bot test Windows x64 Debug Build please
@dotnet-bot test Windows x86 Debug Build please

**Comment by Wraith2** [comment]
@dotnet-bot test Windows x86 Release Build please

**Comment by saurabh500** [comment]
@Wraith2 There are some merge conflicts now. Can you take care of this?

**Comment by Wraith2** [comment]
Done.

**Pull request merged**